numpyro/infer/util.py

```diff
@@ -653,16 +653,17 @@ def initialize_model(
         has_enumerate_support,
         model_trace,
     ) = _get_model_transforms(substituted_model, model_args, model_kwargs)
+
     # substitute param sites from model_trace to model so
     # we don't need to generate again parameters of `numpyro.module`
-    model = substitute(
-        model,
-        data={
-            k: site["value"]
-            for k, site in model_trace.items()
-            if site["type"] in ["param"]
-        },
-    )
+    def substitute_fn(site):
+        if site["type"] in ["param", "mutable"]:
+            return site["value"]
+        elif site["type"] == "prng_key":
+            return random.PRNGKey(0)
+
+    model = substitute(model, substitute_fn=substitute_fn)
+
```

Comment:
let's keep the original model as-is and define a new substitute_fn
```
def substitute_key(msg):
    if msg["type"] == "prng_key":
        return random.PRNGKey(0)

if any(site["type"] == "prng_key" for site in model_trace.values()):
    model = substitute(model, substitute_fn=substitute_key)
```
```
```

Reply:
As in https://github.com/pyro-ppl/numpyro/pull/1823/commits/232c89c08954a19d5578198137f567f3e62081a4? 

When I run the test locally I keep getting ` KeyError: 'name'` coming from 

```python
    def _unconstrain_reparam(params, site):
>       name = site["name"]
E       KeyError: 'name'
E       --------------------
```

I am afraid this will happen everytime we have a `site["name"]` statement.

Reply:
I see. It seems that we can't substitute sites without "name" attribute. How about we define a new handler
```
class _substitute_default_key(Messenger):
    def process_msg(self, msg):
        if msg["type"] == "prng_key" and msg["value"] is None:
            msg["value"] = random.PRNGKey(0)

model = _substitute_default_key(model)

Reply:
@fehiepsi I think this does the trick indeed! :D